File: src/test/java/io/protobase/f7/formulas/math/LOG10Test.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.NumException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class LOG10Test extends TestFormula {
  @Test
  public void testApply() {
    assertThat(LOG10.SELF.apply(null, 128.0)).isEqualTo(Math.log10(128));
    assertThat(LOG10.SELF.apply(null, 1.0)).isEqualTo(Math.log10(1.0));
    assertThat(LOG10.SELF.apply(null, 2.0)).isEqualTo(Math.log10(2.0));
    assertThat(LOG10.SELF.apply(null, 3.0)).isEqualTo(Math.log10(3.0));
    assertThat(LOG10.SELF.apply(null, 10.0)).isEqualTo(Math.log10(10.0));
    assertThat(LOG10.SELF.apply(null, 100.0)).isEqualTo(Math.log10(100.0));
    assertThat(LOG10.SELF.apply(null, 1000.0)).isEqualTo(Math.log10(1000));
    assertThat(LOG10.SELF.apply(null, 10000.0)).isEqualTo(Math.log10(10000));
    assertThat(LOG10.SELF.apply(null, 100000.0)).isEqualTo(Math.log10(100000));
  }

  @Test
  public void testApply_parameterErrors() {
    assertThat(LOG10.SELF.apply(null, 0.0)).isEqualTo(new NumException());
    assertThat(LOG10.SELF.apply(null, -10.1)).isEqualTo(new NumException());
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(LOG10.SELF.apply(null, "10000")).isEqualTo(Math.log10(10000));
  }

  @Test
  public void testApply_grid() {
    assertThat(LOG10.SELF.apply(null,
        Grid.builder().add(0, 0, 44.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(Math.log10(44));
  }

  @Test
  public void test_lookup() {
    LOG10 F = new LOG10(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(128.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(Math.log10(128));
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(LOG10.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(LOG10.SELF.apply(null)).isEqualTo(new NAException());
    assertThat(LOG10.SELF.apply(null, "A", "B")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/LOG10Test.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { LOG10 } from "../../../../main/js/formulas/math/LOG10";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { Converters } from "../../../../main/js/utils/Converters";
import { CommonModels } from "../../testutils/CommonModels";

describe("LOG10", function () {
  it("should work with numbers", function () {
    assert.deepEqual(LOG10.SELF.run(null, 128), Math["log10"](128));
    assert.deepEqual(LOG10.SELF.run(null, 1), Math["log10"](1));
    assert.deepEqual(LOG10.SELF.run(null, 2), Math["log10"](2));
    assert.deepEqual(LOG10.SELF.run(null, 3), Math["log10"](3));
    assert.deepEqual(LOG10.SELF.run(null, 10), Math["log10"](10));
    assert.deepEqual(LOG10.SELF.run(null, 100), Math["log10"](100));
    assert.deepEqual(LOG10.SELF.run(null, 1000), Math["log10"](1000));
    assert.deepEqual(LOG10.SELF.run(null, 10000), Math["log10"](10000));
    assert.deepEqual(LOG10.SELF.run(null, 100000), Math["log10"](100000));
  });

  it("should return errors when parameters are out of bounds", function () {
    assert.deepEqual(
      Converters.castAsF7Exception(LOG10.SELF.run(null, 0)).name,
      F7ExceptionName.NUM
    );
    assert.deepEqual(
      Converters.castAsF7Exception(LOG10.SELF.run(null, -10)).name,
      F7ExceptionName.NUM
    );
  });

  it("should work with strings", function () {
    assert.deepEqual(LOG10.SELF.run(null, "128"), Math["log10"](128));
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(LOG10.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new LOG10(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(128);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), Math["log10"](128));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 128).add(0, 1, "A").build();
    assert.deepEqual(LOG10.SELF.run(null, one), Math["log10"](128));
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((LOG10.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((LOG10.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
